[upd] Document non-denotable types (#5642)
JetBrains/kotlin-web-site: docs/topics/typecasts.md

>
{style="note"}

### Intersection types

When the compiler smart-casts an object through multiple `&&` checks, it infers an [*intersection type*](https://kotlinlang.org/spec/type-system.html#intersection-types).
It is an internal type that simultaneously satisfies all the checked constraints:

```kotlin
interface Bird {
    fun fly()
}

interface Fish {
    fun swim()
}

fun describe(animal: Any) {
    // Infers the Bird and Fish types
    if (animal is Bird && animal is Fish) {
        // Accesses fly() and swim() without additional checks or casts
        animal.fly()
        animal.swim()
    }
}
```

Since intersection types are non-denotable, you can't write them in Kotlin code. However, you may encounter them in

> compiler error messages and IDE tooltips, typically displayed as `A & B`.

Reviewer: Possibly worth to mention the exception of `T & Any`:
```kotlin
fun <T> method(): T & Any = TODO()
```

Author: done

>
{style="note"}

### Intersection types

When the compiler smart-casts an object through multiple `&&` checks, it infers an [*intersection type*](https://kotlinlang.org/spec/type-system.html#intersection-types).

> It is an internal type that simultaneously satisfies all the checked constraints:

Reviewer 2: ```suggestion
This is an internal type that simultaneously satisfies all the checked constraints:
```
Nit: Sounds more natural to a native speaker.

```kotlin
interface Bird {
    fun fly()
}

interface Fish {
    fun swim()
}

fun describe(animal: Any) {
    // Infers the Bird and Fish types
    if (animal is Bird && animal is Fish) {
        // Accesses fly() and swim() without additional checks or casts
        animal.fly()
        animal.swim()
    }
}
```

> Since intersection types are non-denotable, you can't write them in Kotlin code. However, you may encounter them in

Reviewer 2: ```suggestion
Intersection types are non-denotable. They only exist in the compiler's internal type system to preserve type information during type checking. You can't write them directly in Kotlin code. You may encounter intersection types in
```
It's not clear what "non-denotable" means just from this text and I think it's also a fairly advanced English word. What do you think about expanding your explanation here a little? I made a suggestion for you to consider :)

compiler error messages and IDE tooltips, typically displayed as `A & B`.

> The one exception is `T & Any`, used to declare a [definitely non-nullable type](generics.md#definitely-non-nullable-types).

Reviewer 2: Can you change the passive "is used" to something more active? Maybe "which declares"?

> The syntax is reserved specifically to combine a type parameter with `Any`. You still can't write other

Reviewer 2: ```suggestion
This syntax is reserved specifically to combine a type parameter with `Any`. You still can't write other
```

> intersection types explicitly.

Reviewer 2: Do we really need the last sentence since the paragraph starts by explaining that you can't write them in code?

```kotlin

> fun  method(): T & Any = TODO()

Reviewer: It would be great if examples had some meaning
```suggestion
fun  <T> T.assertNotNull(): T & Any = this ?: throw IllegalStateException("null value")
```
